crates/hir-ty/src/method_resolution.rs

```diff
@@ -896,6 +896,10 @@ pub fn iterate_method_candidates_dyn(
             let deref_chain = autoderef_method_receiver(&mut table, ty);
 
             let result = deref_chain.into_iter().try_for_each(|(receiver_ty, adj)| {
+                if matches!(receiver_ty.value.kind(Interner), TyKind::Ref(Mutability::Mut, _, _)) {
+                    // We should prefer reborrow of mutable references instead of moving them.
+                    return ControlFlow::Continue(());
+                }
```

Comment:
Can we do this directly in `autoderef_method_receiver`? That way we a) allocate less, b) indexing also benefits from this which I would expect should share this behavior (if not then ignore this comment, haven't checked how rustc does it)?

Reply:
`autoderef_method_receiver` is already emitting deref for mutable references, but `iterate_method_candidates_with_autoref` sees that the method can accept the receiver without any deref, and returns `Break` so the rest of things `autoderef_method_receiver` returned is ignored. It is probably possible to make `autoderef_method_receiver` lazy, but that's another topic.

About index, this is not concern there, since we always emit `&mut` or `&`, so reborrow always happen even for immutable reference.

But I noticed that the current implementation is wrong. Let me fix it.